### PR TITLE
Panic with GitLab project repository auth

### LINKED ISSUE #1111: Panic with GitLab project auth when user has no access
When using GitLab as a provide and setting the option `--gitlab-project=ska-telescope/ska-tango-images=30` there is a panic caused at https://github.com/oauth2-proxy/oauth2-proxy/blob/master/providers/gitlab.go#L334 when the user has no access as `perms` is a `nil` value at this point.

## Expected Behavior

No panic cause - user gets the expected 403 screen as they have no access.

## Current Behavior

User gets a 502 gateway error. oauth2-proxy panics:
```
2021/03/18 06:58:01 http: panic serving 127.0.0.1:45824: runtime error: invalid memory address or nil pointer dereference                                                    
goroutine 568 [running]:                                                                                                                              
net/http.(*conn).serve.func1(0xc000124be0)                                                                                                                                
        /usr/local/go/src/net/http/server.go:1824 +0x153                                                                                                                    
panic(0xc84f20, 0x135ed50)                                                                                                                                                   
        /usr/local/go/src/runtime/panic.go:971 +0x499                                                                                                                        
github.com/oauth2-proxy/oauth2-proxy/v7/providers.(*GitLabProvider).addProjectsToSession(0xc00020fd00, 0xe89098, 0xc000507140, 0xc00018e480)
        /go/src/github.com/oauth2-proxy/oauth2-proxy/providers/gitlab.go:335 +0x154                                                                                          
github.com/oauth2-proxy/oauth2-proxy/v7/providers.(*GitLabProvider).EnrichSession(0xc00020fd00, 0xe89098, 0xc000507140, 0xc00018e480, 0x1, 0x0)                           
        /go/src/github.com/oauth2-proxy/oauth2-proxy/providers/gitlab.go:301 +0x174                                                                                          
main.(*OAuthProxy).enrichSessionState(0xc000142900, 0xe89098, 0xc000507140, 0xc00018e480, 0x0, 0x9)                                                                          
        /go/src/github.com/oauth2-proxy/oauth2-proxy/oauthproxy.go:882 +0x63                                                       
```

## Possible Solution

```
diff --git a/providers/gitlab.go b/providers/gitlab.go
index f54430f..d0e090c 100644
--- a/providers/gitlab.go
+++ b/providers/gitlab.go
@@ -331,7 +331,7 @@ func (p *GitLabProvider) addProjectsToSession(ctx context.Context, s *sessions.S
                                perms = projectInfo.Permissions.GroupAccess
                        }
 
-                       if perms.AccessLevel >= project.AccessLevel {
+                       if perms != nil && perms.AccessLevel >= project.AccessLevel {
                                s.Groups = append(s.Groups, fmt.Sprintf("project:%s", project.Name))
                        } else {
                                logger.Errorf("Warning: user %q does not have the minimum required access level for project %q", s.Email, project.Name)
```

## Steps to Reproduce (for bugs)

Configure with an options like `--gitlab-project=ska-telescope/ska-tango-images=30`, and try to login with a user that has no access.

## Your Environment

- Version used:  affects current head and v7.

### PR DESCRIPTION
**Pull request by piersharding**
* /api/v4/projects/:id can return nil permissions

Signed-off-by: Piers Harding <piers@ompka.net>

## Description

Check project API payload for `nil` in GroupAccess value.

fixes #1111 

## Motivation and Context

This PR addresses #1111 - where the GitLab projects API (https://docs.gitlab.com/ee/api/projects.html) does not necessarily return a GroupAccess value.  The resulting permissions are now checked for `nil` to mitigate this.

## How Has This Been Tested?

The change has been tested with setting `--gitlab-project=ska-telescope/ska-tango-images=30`, and running against the patched image registry.gitlab.com/piersharding/oauth2-proxy:latest .

## Checklist:

(not too sure about these settings)

- [ ] My change requires a change to the documentation or CHANGELOG.
- [ ] I have updated the documentation/CHANGELOG accordingly.
- [ ] I have created a feature (non-master) branch for my PR.

### CODE REVIEW

File: providers/gitlab.go [hunk 1]
```diff
@@ -331,7 +331,7 @@ func (p *GitLabProvider) addProjectsToSession(ctx context.Context, s *sessions.S
 				perms = projectInfo.Permissions.GroupAccess
 			}
 
-			if perms.AccessLevel >= project.AccessLevel {
+			if perms != nil && perms.AccessLevel >= project.AccessLevel {
```
Comment (NickMeves) [inline providers/gitlab.go:334]: This mixed with the above `if perms == nil` fallback handling is slightly confusing. Can we check for `nil` on line 331 on `projectInfo.Permissions.GroupAccess` and `continue` if it is?

Can you also add a scenario for this case to the test table here: https://github.com/oauth2-proxy/oauth2-proxy/blob/4d9de06b1dbad8d27965afa2586df29667eeb50c/providers/gitlab_test.go#L189

Comment (piersharding) [reply providers/gitlab.go:339]: Thanks - I have reformatted it as suggested.

**Comment by papey** [comment]
@piersharding can you also add a test case for this ? Thank !

**Review by NickMeves — APPROVED** [review]
Thanks for the fix!

**Comment by NickMeves** [comment]
We are releasing in a couple of hours, we'll take care of adding a CHANGELOG entry in the release PR.

**Comment by piersharding** [comment]
Thanks - sorry, I was learning the process as I was going along.
Cheers,
Piers.

**Comment by NickMeves** [comment]
No worries! Thanks for the finding the issue & submitting the fix!

**Pull request merged**